File: pylox/scanner.py
import logging

from pylox.tokens import Token, TokenType, KEYWORDS

logger = logging.getLogger("pylox.scanner")


class Scanner(object):
    def __init__(self, source):
        self.source = source
        self.start = 0
        self.current = 0
        self.line = 1
        self.tokens = []

    def scan_tokens(self):
        """Parses all tokens until EOF for the provided source code string."""
        while not self._is_at_end():
            self.start = self.current
            self.scan_token()
        self.tokens.append(Token(TokenType.EOF, "", None, self.line))
        return self.tokens

    def _is_at_end(self):
        return self.current >= len(self.source)

    def _add_token(self, token_type, literal=None):
        text = self.source[self.start:self.current]
        self.tokens.append(Token(token_type, text, literal, self.line))

    def _advance(self):
        self.current += 1
        return self.source[self.current - 1]

    def _match(self, expected):
        if self._is_at_end() or self.source[self.current] != expected:
            return False
        self.current += 1
        return True

    def _peek(self):
        return '\0' if self._is_at_end() else self.source[self.current]

    def _peek_next(self):
        index = self.current + 1
        return self.source[index] if index < len(self.source) else '\0'

    def scan_token(self):
        c = self._advance()
        if c == '(':
            self._add_token(TokenType.LEFT_PAREN)
        elif c == ')':
            self._add_token(TokenType.RIGHT_PAREN)
        elif c == '{':
            self._add_token(TokenType.LEFT_BRACE)
        elif c == '}':
            self._add_token(TokenType.RIGHT_BRACE)
        elif c == ',':
            self._add_token(TokenType.COMMA)
        elif c == '.':
            self._add_token(TokenType.DOT)
        elif c == '-':
            self._add_token(TokenType.MINUS)
        elif c == '+':
            self._add_token(TokenType.PLUS)
        elif c == ';':
            self._add_token(TokenType.SEMICOLON)
        elif c == '*':
            self._add_token(TokenType.STAR)
        elif c == '!':
            self._add_token(TokenType.BANG_EQUAL if self._match("=") else TokenType.BANG)
        elif c == '=':
            self._add_token(TokenType.EQUAL_EQUAL if self._match("=") else TokenType.EQUAL)
        elif c == '<':
            self._add_token(TokenType.LESS_EQUAL if self._match("=") else TokenType.LESS)
        elif c == '>':
            self._add_token(TokenType.GREATER_EQUAL if self._match("=") else TokenType.GREATER)
        elif c == '/':
            if self._match('/'):
                # comments go the end of the line
                while self._peek() != '\n' and not self._is_at_end():
                    self._advance()
            else:
                self._add_token(TokenType.SLASH)
        elif c in ' \r\t':
            # ignore whitespace
            pass
        elif c == '\n':
            self.line += 1
        elif c == '"':
            self._scan_string()
        elif self._is_digit(c):
            self._scan_number()
        elif self._is_alpha(c):
            self._scan_identifier()
        else:
            logger.error("Unexpected character: '{}'".format(c))

    def _scan_string(self):
        while self._peek() != '"' and not self._is_at_end():
            if self._peek() == '\n':
                # multiline string!
                self.line += 1
            self._advance()

        if self._is_at_end():
            # unterminated string!
            print("ERROR! unterminated string!")
            return

        # move past close of string
        self._advance()
        text = self.source[self.start + 1:self.current - 1]
        self._add_token(TokenType.STRING, text)

    def _is_digit(self, c):
        return c >= '0' and c <= '9'

    def _is_alpha(self, c):
        return (c >= 'a' and c <= 'z') or (c >= 'A' and c <= 'Z') or c == '_'

    def _is_alphanumeric(self, c):
        return self._is_digit(c) or self._is_alpha(c)

    def _scan_number(self):
        while self._is_digit(self._peek()):
            self._advance()
        if self._peek() == '.' and self._is_digit(self._peek_next()):
            self._advance()
        while self._is_digit(self._peek()):
            self._advance()
        self._add_token(TokenType.NUMBER, float(self.source[self.start:self.current]))

    def _scan_identifier(self):
        while self._is_alphanumeric(self._peek()):
            self._advance()

        text = self.source[self.start:self.current]
        if text in KEYWORDS:
            # pylox-defined keyword
            self._add_token(KEYWORDS[text])
        else:
            # user-defined keyword
            self._add_token(TokenType.IDENTIFIER)

File: klox/src/main/kotlin/scanner.kt
package com.mab.lox

import com.mab.lox.TokenType.*

class Scanner(private val source: String) {
    private val tokens = ArrayList<Token>()
    private var start = 0
    private var current = 0
    private var line = 1

    private val keywords = mapOf(
        "and" to AND, "class" to CLASS, "else" to ELSE, "false" to FALSE, "for" to FOR, "fun" to FUN, "if" to IF,
        "nil" to NIL, "or" to OR, "print" to PRINT, "return" to RETURN, "super" to SUPER, "this" to THIS,
        "true" to TRUE, "var" to VAR, "while" to WHILE
    )

    private fun isAtEnd() = current >= source.length

    private fun isAlpha(c: Char) = (c in 'a'..'z') || (c in 'A'..'Z') || c == '_'

    private fun isDigit(c: Char) = c in '0'..'9'

    private fun isAlphaNumeric(c: Char) = isAlpha(c) || isDigit(c)

    private fun peek(): Char {
        if (isAtEnd()) return 0.toChar()
        return source[current]
    }

    private fun peekNext(): Char {
        if (current + 1 >= source.length) return 0.toChar()
        return source[current + 1]
    }

    private fun match(expected: Char): Boolean {
        if (isAtEnd()) return false
        if (source[current] != expected) return false

        // Matched expected. Consume next character.
        advance()
        return true
    }

    private fun advance(): Char {
        return source[current++]
    }

    private fun addToken(type: TokenType, literal: Any? = null) {
        val text = source.substring(start, current)
        tokens.add(Token(type, text, literal, line))
    }

    fun scanTokens(): List<Token> {
        while (!isAtEnd()) {
            start = current
            scanToken()
        }
        tokens.add(Token(EOF, "", null, line))
        return tokens
    }

    private fun scanToken() {
        when (val c = advance()) {
            '(' -> addToken(LEFT_PAREN)
            ')' -> addToken(RIGHT_PAREN)
            '{' -> addToken(LEFT_BRACE)
            '}' -> addToken(RIGHT_BRACE)
            ',' -> addToken(COMMA)
            '.' -> addToken(DOT)
            '-' -> addToken(MINUS)
            '+' -> addToken(PLUS)
            ';' -> addToken(SEMICOLON)
            '*' -> addToken(STAR)
            '!' -> addToken(if (match('=')) BANG_EQUAL else BANG)
            '=' -> addToken(if (match('=')) EQUAL_EQUAL else EQUAL)
            '<' -> addToken(if (match('=')) LESS_EQUAL else LESS)
            '>' -> addToken(if (match('=')) GREATER_EQUAL else GREATER)
            '/' -> {
                if (match('/')) {
                    // A comment goes until the end of the line.
                    while (peek() != '\n' && !isAtEnd()) advance()
                } else {
                    addToken(SLASH)
                }
            }
            ' ', '\r', '\t' -> {} // Ignore whitespace.
            '\n' -> line++
            '"' -> string()
            else -> {
                if (isDigit(c)) {
                    number()
                } else if (isAlpha(c)) {
                    identifier()
                } else {
                    Lox.error(line, "Unexpected character.")
                }

            }
        }
    }

    /**
     * `STRING -> "\"" <any char except "\"">* "\"";`
     */
    private fun string() {
        // Handler for parsing remaining portion of string.
        while (peek() != '"' && !isAtEnd()) {
            // Ensure we count lines as we're reading strings.
            if (peek() == '\n') line++
            // Continue reading the string.
            advance()
        }

        if (isAtEnd()) {
            Lox.error(line, "Unterminated string.")
            return
        }

        // Move past the closing ".
        advance()

        // Trim the surrounding quotes.
        val value = source.substring(start + 1, current - 1)
        addToken(STRING, value)
    }

    /**
     * `IDENTIFIER -> ALPHA ( ALPHA | DIGIT )* ;`
     */
    private fun identifier() {
        while (isAlphaNumeric(peek())) advance()

        val text = source.substring(start, current)
        addToken(keywords.getOrDefault(text, IDENTIFIER))
    }

    /**
     * `NUMBER -> DIGIT+ ( "." DIGIT+ )? ;`
     */
    private fun number() {
        while (isDigit(peek())) advance()

        val isInt = peek() != '.'
        if (peek() == '.' && isDigit(peekNext())) {
            // Consume the '.'
            advance()
            // Parse remaining digits.
            while (isDigit(peek())) advance()
        }

        val literalStr = source.substring(start, current)
        val literal = if (isInt) literalStr.toInt() else literalStr.toDouble()
        addToken(NUMBER, literal)
    }
}
